akka-http-core/src/main/scala/akka/http/engine/client/ConnectionPool.scala

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.engine.client
+
+import java.net.InetSocketAddress
+import scala.util.{ Success, Try }
+import scala.concurrent.{ Promise, Future }
+import akka.event.LoggingAdapter
+import akka.actor.{ ActorSystem, Props, ActorRef }
+import akka.stream.{ FlattenStrategy, FanOutShape2, FlowShape, FlowMaterializer }
+import akka.stream.impl.ActorProcessor
+import akka.stream.scaladsl._
+import akka.http.model._
+import akka.http.util._
+import akka.http.Http
+
+private[http] object ConnectionPool {
+  import HostConnectionPoolGateway._
+
+  type ClientFlow = Flow[(HttpRequest, Any), (Try[HttpResponse], Any), Http.HostConnectionPool]
+
+  final class HostPool(flow: ClientFlow, val isShutdown: Future[Unit], val doShutdown: () ⇒ Future[Unit]) {
+    def clientFlow[T] = flow.asInstanceOf[Flow[(HttpRequest, T), (Try[HttpResponse], T), Http.HostConnectionPool]]
+  }
+
+  val CowboyEntry = new HostPool(null, null, null) // for reigning in potentially approaching thundering herds
+
+  private val poolClientActorName = new SeqActorName("PoolClient")
+  private val poolGatwayActorName = new SeqActorName("PoolGateway")
+
+  /**
+   * Creates a new "super pool flow", which routes incoming requests to "their" host connection pool,
+   * depending on the respective effective URI.
+   */
+  def superPool[T](setup: ConnectionPoolSetup)(implicit system: ActorSystem, fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), Unit] = {
+    Flow.wrap {
+      FlowGraph.partial() { implicit b ⇒
+        import FlowGraph.Implicits._
+        val map = b.add(Flow[(HttpRequest, T)].map {
+          case (r, ctx) ⇒ Try(r.withEffectiveUri(securedConnection = false)) -> ctx
+        })
+        val failureSplit = b.add(new FailureSplit[T])
+        val mainFlow = Flow[(HttpRequest, T)]
+          .groupBy { case (req, _) ⇒ req.uri.authority }
+          .map {
+            case (Uri.Authority(host, port, _), hostRequests) ⇒
+              val poolClientFlow = Http().cachedHostConnectionPool[T](HostConnectionPoolSetup(host.toString(), port, setup))
+              hostRequests.via(poolClientFlow)
+          }
+          .flatten(FlattenStrategy.Concat()) // WRONG! FIXME: Needs to be `FlattenStrategy.Merge` instead! (#15089)
```

Comment:
This might in theory work, but it is a rather costly way of formulating this: as we discussed this would be better off using `mapAsync` and using a single-request–response protocol. Trying to amortize the stream materialization costs will lead into big troubles (idle timeouts shutting things down, which is nasty to get to work consistently; maximum number of open groups in this expression; both connected to double-bookkeeping here and in cachedHostConnectionPool).